lib/spack/spack/test/spec_yaml.py

```diff
@@ -327,9 +328,71 @@ def test_save_dependency_spec_yamls_subset(tmpdir, config):
         spec_a.concretize()
         b_spec = spec_a['b']
         c_spec = spec_a['c']
-        spec_a_yaml = spec_a.to_yaml(hash=ht.build_hash)
-
-        save_dependency_spec_yamls(spec_a_yaml, output_path, ['b', 'c'])
-
-        assert check_specs_equal(b_spec, os.path.join(output_path, 'b.yaml'))
-        assert check_specs_equal(c_spec, os.path.join(output_path, 'c.yaml'))
+        spec_a_json = spec_a.to_json(hash=ht.build_hash)
+
+        save_dependency_specfiles(spec_a_json, output_path, ['b', 'c'])
+
+        assert check_specs_equal(b_spec, os.path.join(output_path, 'b.json'))
+        assert check_specs_equal(c_spec, os.path.join(output_path, 'c.json'))
+
+
+def test_legacy_yaml(tmpdir, install_mockery, mock_packages):
+    """Tests a simple legacy YAML with a dependency and ensures spec survives
+    concretization."""
+    yaml = """
+spec:
+- a:
+    version: '2.0'
+    arch:
+      platform: linux
+      platform_os: rhel7
+      target: x86_64
+    compiler:
+      name: gcc
+      version: 8.3.0
+    namespace: builtin.mock
+    parameters:
+      bvv: true
+      foo:
+      - bar
+      foobar: bar
+      cflags: []
+      cppflags: []
+      cxxflags: []
+      fflags: []
+      ldflags: []
+      ldlibs: []
+    dependencies:
+      b:
+        hash: iaapywazxgetn6gfv2cfba353qzzqvhn
+        type:
+        - build
+        - link
+    hash: obokmcsn3hljztrmctbscmqjs3xclazz
+    full_hash: avrk2tqsnzxeabmxa6r776uq7qbpeufv
+    build_hash: obokmcsn3hljztrmctbscmqjs3xclazy
+- b:
+    version: '1.0'
+    arch:
+      platform: linux
+      platform_os: rhel7
+      target: x86_64
+    compiler:
+      name: gcc
+      version: 8.3.0
+    namespace: builtin.mock
+    parameters:
+      cflags: []
+      cppflags: []
+      cxxflags: []
+      fflags: []
+      ldflags: []
+      ldlibs: []
+    hash: iaapywazxgetn6gfv2cfba353qzzqvhn
+    full_hash: qvsxvlmjaothtpjluqijv7qfnni3kyyg
+    build_hash: iaapywazxgetn6gfv2cfba353qzzqvhy
+"""
+    spec = Spec.from_yaml(yaml)
+    non_concrete_spec = spec.copy()
+    spec.concretize()
+    assert non_concrete_spec.eq_dag(spec)
```

Comment:
We should avoid calling it "non_concrete", since the spec as read in will in fact be concrete (that's what we confirm in the test).

```suggestion
    spec = Spec.from_yaml(yaml)
    concrete_spec = spec.concretized()
    assert concrete_spec.eq_dag(spec)
```